libsolidity/ast/Types.cpp

```diff
@@ -3666,12 +3669,26 @@ FunctionTypePointer FunctionType::asExternallyCallableFunction(bool _inLibrary)
 {
 	TypePointers parameterTypes;
 	for (auto const& t: m_parameterTypes)
+	{
+		// TODO :: Add warning/error here for using calldata arguments
+		// solAssert(
+		// 	!TypeProvider::isReferenceWithLocation(t, DataLocation::CallData),
+		// 	"Calldata arguments not allowed for externally callable functions. Use memory arguments instead."
+		// );
 		if (TypeProvider::isReferenceWithLocation(t, DataLocation::CallData))
+		{
+			m_errorReporter.warning(
```

Comment:
I am not sure this is the right place to report the warning or "hint" to the user. 
Since this problem happens when converting from external functions with `calldata` arguments (which are altered here to `memory`), maybe it is necessary to identify the situation  in the context of `TypeChecker`.

Reply:
`asExternallyCallableFunction` is called from different places in the code for the different test cases that we're adding for this issue.

Instead of checking for `external` functions having `calldata` arguments around each of these instances, I thought it prudent to add the warning in this function itself. 

This way, we can be sure that, whenever we are trying to get an `external` function, that has had `calldata `arguments, which will now be converted to `memory` arguments, we can print out a warning.

Reply:
Perhaps, we could even disallow it , as @chriseth  mentions in https://github.com/ethereum/solidity/issues/12778#issuecomment-1169917941 

@cameel  any thoughts ?

Reply:
> This way, we can be sure that, whenever we are trying to get an `external` function, that has had `calldata `arguments, which will now be converted to `memory` arguments, we can print out a warning. 

The problem is that `asExternallyCallableFunction()` may be called in different contexts, not just during analysis. For example in the codegen where we don't even have access to an error reporter and nowhere really to print the error (you can print it to the console but what if the user requested StandardJSON?). I see that you actually added a reporter to `FunctionType` and I don't think that matches the current design of the type classes.

Another downside is that we can now get this warning multiple times - each time the function happens to be called anywhere. We only really want the warning when we are processing that part of the code during the analysis. Later we don't care. Best if we could have an assert that will somehow detect that we didn't catch this in the analysis but I don't think it's doable if it's just a warning and not an error.

Reply:
I've gone through the code. And by my judgement, it should be okay inserting a warning around `ContractLevelChecker.cpp::checkExternalTypeClashes(ContractDefinition const& _contract)`

The `asExternallyCallableFunction` call here is being made for all test cases. And the function itself seems to be part of the analysis phase. 

@cameel  thoughts on this ?

Reply:
I consider this a better location to report the warning. Although I guess that some unrelated situations will get the warning too.

Reply:
But how would that work? That code is processing all function definitions in all contracts. I don't think we want to warn about them having `calldata` arguments, do we? Or am I misunderstanding something here?

I think this check should go into some place that visits expressions like assignments, passing in parameters, etc.

Reply:
I think 

> But how would that work? That code is processing all function definitions in all contracts. I don't think we want to warn about them having `calldata` arguments, do we? Or am I misunderstanding something here?
> I think this check should go into some place that visits expressions like assignments, passing in parameters, etc.

I think that I was confused and not sure about the objective. Reading the discussion again, I think that we want to hint the user only when there is the type mismatch error involving pointers of function with `calldata` arguments, right? 
That is what I had understood initially and hence my [suggestion](https://github.com/ethereum/solidity/pull/13227#pullrequestreview-1065866454) earlier, pointing out to method`TypeChecker::visit(VariableDeclarationStatement const&)`.

What do you think about this @nishant-sachdeva ?

Reply:
> I think that we want to hint the user only when there is the type mismatch error involving pointers of function with `calldata` arguments, right?

Yes, that's correct. `TypeChecker::visit(VariableDeclarationStatement const&)` might be one of the places where we want to do this but I don't think it's the only one. You can have assignments in expressions, you can pass parameters to functions, etc. and at all those places the conversion will happen.

Actually would be nice if we could do it in a more general way like @nishant-sachdeva tried to, but that still should be in the type checker, not in type classes. Maybe we have some helper we use for type checking conversions? If not, maybe we could create one. But hard to say if that's feasible without actually going through the code.

Reply:
@matheusaaguiar yes, the initial problem was 

> 

From what I understand, this is a symptom of a larger issue that calldata arguments are not allowed for external functions. The user should, in my opinion, be warned about using calldata args in external functions. Hence I was looking for a common place where all the external functions/ arguments are processed and add a warning there.

Reply:
> I think the user should be warned about using calldata args in external functions because it can create problems other than function pointer assignments as well.

Ah, there seems to be a misunderstanding. You're probably thinking about `calldata` args in **function pointers** not **function definitions**? `calldata` args in external functions are actually recommended whenever possible. It's just the pointers where they might be an issue.

So far you've been pointing at places that process function definitions, not pointers. Same with `TypeChecker::visit(FunctionDefinition const& _function)` - it goes over definitions. You'd need to go through variable declarations instead. `TypeChecker::typeCheckFunctionCall` on the other hand does not seem relevant. We're interested in function pointers whether they're being called or not.

But overall that's a different direction than I thought we were going. Eventually we may want to even disallow such pointers (suggested that in https://github.com/ethereum/solidity/issues/12778#issuecomment-1169917941) but that discussion is not yet resolved. I actually did not consider simply warning about function pointers taking `calldata` args and this may be a good idea, but it's a design decision we should discuss first (you could add that to the agenda for today's call). For the time being I was assuming that the goal in this PR was just to adjust the error message, which happens when function pointers are passed around or assigned.

Reply:
Indeed, the errors we are seeing in our test cases are at the variable assignment stage. 

But the `calldata` args of the `external function pointers` have already been converted to `memory` args. I'm trying to locate the place where that happens. I think, inserting a warning / error ( should we choose to disallow it completely ) there should be enough to close this issue.

Reply:
> We actually still have this to discuss in the agenda. Judging from what was said last time, I think it's likely that we will want to disallow such pointers after all. Or do something else that will make this warning irrelevant.
My suggestion would be to just keep the docs and test cases for now and leave the error message tweak out of the PR completely. Just post in the issue that the warning part was not done.

Closing this thread for now. Will resolve the other comments and look to wrap up this PR